planner/core/expression_rewriter.go

```diff
@@ -1237,7 +1237,7 @@ func (er *expressionRewriter) patternLikeToExpression(v *ast.PatternLikeExpr) {
 		}
 		if !isNull {
 			patValue, patTypes := stringutil.CompilePattern(patString, v.Escape)
-			if stringutil.IsExactMatch(patTypes) {
+			if stringutil.IsExactMatch(patTypes) && er.ctxStack[l-2].GetType().EvalType().IsStringKind() {
```

Comment:
Is it seems that we should not call `IsStringKind()` but check `.EvalType() == ETString` here?

Reply:
> Is it seems that we should not call `IsStringKind()` but check `.EvalType() == ETString` here?

It seems that IsStringKind() is enough, If the val's type is String, they will be compared as string.
https://github.com/pingcap/tidb/blob/5aef053c163524aa167e6bb4ddabd0671a75e712/expression/builtin_compare.go#L958-L978

Reply:
IsStringKind contains ETDatetime, etc.

``` sql
func (et EvalType) IsStringKind() bool {
	return et == ETString || et == ETDatetime ||
		et == ETTimestamp || et == ETDuration || et == ETJson
}
```
> If the val's type is String, they will be compared as string

Noop, `getBaseCmpType` gets a *basic* comparison type, the final comparison type may change later.

Reply:
Though not related to this patch, looks like I found another compatibility issue:
In MySQL:
```
mysql> create table tbl(id int primary key, a date);
Query OK, 0 rows affected (0.03 sec)

mysql> set sql_mode = '';
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> insert tbl values(0, '0000-00-00');
Query OK, 1 row affected (0.02 sec)

mysql> select a like 'a' from tbl;
+------------+
| a like 'a' |
+------------+
|          0 |
+------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+---------------------------------------------------+
| Level   | Code | Message                                           |
+---------+------+---------------------------------------------------+
| Warning | 1292 | Incorrect date value: 'a' for column 'a' at row 1 |
+---------+------+---------------------------------------------------+
```
while in TiDB (even with this patch):
```
mysql> select a like 'a' from tbl;
+------------+
| a like 'a' |
+------------+
|       NULL |
+------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-------------------------------+
| Level   | Code | Message                       |
+---------+------+-------------------------------+
| Warning | 1292 | Incorrect datetime value: 'a' |
+---------+------+-------------------------------+
```
@wjhuang2016 are you interested to fix this as well?

Reply:
> IsStringKind contains ETDatetime, etc.
> 
> ```sql
> func (et EvalType) IsStringKind() bool {
> 	return et == ETString || et == ETDatetime ||
> 		et == ETTimestamp || et == ETDuration || et == ETJson
> }
> ```
> 
> > If the val's type is String, they will be compared as string
> 
> Noop, `getBaseCmpType` gets a _basic_ comparison type, the final comparison type may change later.

I got it. If the lhs is a datatime, the rhs may be converted to a datatime.